### PR TITLE
Fix tests on windows

### LINKED ISSUE #632: Tests do not pass on windows
Under windows some tests of alembic fail.

The main reason for failures are 
- a Permission error `PermissionError: [WinError 32] The process cannot access the file because it is being used by another process: 'scratch\\scripts\\foo.db'`, when setting up some tests with sqlite
- slash/backslash difference ` call('foobar\\versions') != call('foobar/versions')`
- newline with `\n` vs `\r\n`

This is mainly to let you know, I'm not sure if it feasible to ensure that the tests work also on windows, since I guess it's hard to setup a ci environment with windows

if this is the expect behavior it can be closed

### PR DESCRIPTION
**Pull request by CaselIT**
Update the tests to make them run on windows
### Description

The main changes are:
- Correctly manage the paths on windows vs *nix like system.
- Manage the new lines \r\n vs \n
- Windows cannot remove open files, namely the sqlite db, so it keeps track of the connections and closes them before removing the files.

I've tested with sqlite, postgres and mysql and all pass (on windows 10)

This fixes #632 
### Checklist

This pull request is:

- [ ] A documentation / typographical error fix
	- Good to go, no issue or tests are needed
- [x] A short code fix
	- please include the issue number, and create an issue if none exists, which
	  must include a complete example of the issue.  one line code fixes without an
	  issue and demonstration will not be accepted.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.   one line code fixes without tests will not be accepted.
- [ ] A new feature implementation
	- please include the issue number, and create an issue if none exists, which must
	  include a complete example of how the feature would look.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.

**Have a nice day!**

### CODE REVIEW

File: alembic/testing/env.py [hunk 1]
```diff
@@ -90,10 +87,53 @@ def env_file_fixture(txt):
...
+        _engine_registry[url] = entry
+        _orig_connect = eng.connect
+
+        def connect(*args, **kwargs):
+            conn = _orig_connect(*args, **kwargs)
+            entry["connections"].append(conn)
+            return conn
+
+        eng.connect = connect
+        return eng
```
Comment (CaselIT) [inline alembic/testing/env.py]: I'm not sure if there is a more elegant way to get all the connections. This version seems to work but it less than elegant

Comment (zzzeek) [reply alembic/testing/env.py]: oh.   right, so yeah, there's no need to monkeypatch, there's an event system that intercepts these.    within the testing suite, this should already be happening, that is, if you call:

    from sqlalchemy.testing import engines
    engines.testing_reaper.close_all()

this should be getting called after every test suite teardown.     it's getting called right here:

https://github.com/sqlalchemy/alembic/blob/master/alembic/testing/plugin/plugin_base.py#L28

so basically none of the above should be needed.

Comment (zzzeek) [reply alembic/testing/env.py]: by which I mean, if we are getting a stack trace there and it is failing to remove the files, some test or fixture is not correctly using the testing.engines package to track that engine.

Comment (CaselIT) [reply alembic/testing/env.py]: The stack trace is on "foo.db". Cannot remove open file.

The stack trace happens on basically every test that uses the db when using sqlite

Comment (CaselIT) [reply alembic/testing/env.py]: One of the classes that has the error is this one `tests/test_command.py::StampMultipleHeadsTest`

It calls only `clear_staging_env` in its tear down

> this should be getting called after every test suite teardown. it's getting called right here: 

this seems to only work once at the end of the class, not after every test.
So it seems that `clear_staging_env` runs after every test, but the engine clear runs only at the end of the class.

I'll use `engines.testing_reaper.close_all()` in `clear_staging_env`

**Comment by CaselIT** [comment]
I've updated the patch to use `engines.testing_reaper.close_all()`

File: alembic/testing/util.py [hunk 2]
```diff
@@ -78,3 +80,11 @@ def run_ddl(self):
         return fixture_functions.fixture(scope=ddl)(run_ddl)
 
     return decorate
+
+
+def path(linuxPath):
```
Comment (zzzeek) [inline alembic/testing/util.py]: oops no camels allowed :)

flake8 will probably hit this

File: alembic/testing/env.py [hunk 3]
```diff
@@ -60,6 +60,9 @@ def staging_env(create=True, template="generic", sourceless=False):
 
 
 def clear_staging_env():
+    from sqlalchemy.testing import engines
+
+    engines.testing_reaper.close_all()
```
Comment (zzzeek) [inline alembic/testing/env.py]: greeeaaat!!!!!

File: alembic/testing/util.py [hunk 4]
```diff
@@ -78,3 +80,11 @@ def run_ddl(self):
...
+    """Takes a linux path with / and transforms it to the correct os path.
+    This is mainly to support windows \\"""
+    if win:
+        return join(*linuxPath.split("/"))
```
Comment (zzzeek) [inline alembic/testing/util.py]: doesnt.....os.path.normpath do this?

File: tests/test_command.py [hunk 5]
```diff
@@ -95,14 +96,17 @@ def _eq_cmd_output(self, buf, expected, env_token=False, currents=()):
...
+            entry = rev.log_entry
+            if os.linesep != '\n':
+                entry = entry.replace('\n', os.linesep)
+            assert_lines.append(entry)
 
         if env_token:
             assert_lines.insert(0, "environment included OK")
 
         eq_(
             buf.getvalue().decode("ascii", "replace").strip(),
```
Comment (zzzeek) [inline tests/test_command.py]: so on this side we are taking "raw" content and converting to something agnostic, so let's make it:

    buf.getvalue().decode("ascii", "replace").replace(os.linesep, "\n").strip()    

and that's it, then we don't need a conditonal above or anything

File: tests/test_command.py [hunk 6]
```diff
@@ -1113,7 +1117,7 @@ def access_(path, mode):
             command.init(self.cfg, directory="foobar")
             eq_(
                 makedirs.mock_calls,
-                [mock.call("foobar"), mock.call("foobar/versions")],
+                [mock.call("foobar"), mock.call(path("foobar/versions"))],
```
Comment (zzzeek) [inline tests/test_command.py]: can we call "path" something more descriptive like "ospath("/foo/bar")"

File: tests/test_script_production.py [hunk 7]
```diff
@@ -1248,7 +1248,9 @@ def test_env_emits_warning(self):
                     )
                 )
```
Comment (zzzeek) [inline tests/test_script_production.py]: would we...do the replace("\\", ..) on the path part here and that's it?  I'm not understanding this escaping should we just make it an r"" string?

File: alembic/testing/util.py (hunk 2 above)
Comment (CaselIT) [reply alembic/testing/util.py]: I usually code in camels so sometimes I use them out of habit

File: alembic/testing/util.py (hunk 4 above)
Comment (CaselIT) [reply alembic/testing/util.py]: Indeed, I remembered there was something that did this, but I did not remember what it was.

File: tests/test_command.py (hunk 5 above)
Comment (CaselIT) [reply tests/test_command.py]: I preferred to change the expected value, but in this case I guess it's best change the actual value

File: tests/test_script_production.py (hunk 7 above)
Comment (CaselIT) [reply tests/test_script_production.py]: the problem is that `assertions.expect_warnings(mgs)` transforms `mgs` (typo.. 😞) to a regexp, so that causes problem.

So we have something like this

```py
msg = '''someting something
C:\\foo\\bar\\path
foo bar
'''
# in expect_warnings
re.compile(msg)
```

this does not work, since we an error like `error: bad escape \p at position 29 (line 2, column 11)`

Comment (CaselIT) [reply tests/test_script_production.py]: I have not checked if we can disable the regexp parse in `expect_warnings`. Maybe that's the best solution

Comment (zzzeek) [reply tests/test_script_production.py]: oh it's a reg.   OK...maybe add an expect_warnings_plain() function then.   this test is manufacturing the assertions from the data which is kind of backwards.

**Comment by CaselIT** [comment]
I think I've updated all review comments

File: tests/test_script_production.py [hunk 8]
```diff
@@ -1248,7 +1248,8 @@ def test_env_emits_warning(self):
...
             )
-        ):
+        )
+        with assertions.expect_warnings(msg, regex=False):
```
Comment (zzzeek) [inline tests/test_script_production.py]: oh looky that I already had it, neat

File: alembic/testing/util.py [hunk 9]
```diff
@@ -78,3 +80,11 @@ def run_ddl(self):
         return fixture_functions.fixture(scope=ddl)(run_ddl)
 
     return decorate
+
+
+def ospath(linux_path):
+    """Takes a linux path with / and transforms it to the correct os path.
+    This is mainly to support windows \\"""
+    if win:
+        return normpath(linux_path)
```
Comment (zzzeek) [inline alembic/testing/util.py]: could we just be calling os.path.normpath in all cases, directly?    e.g. we don't need the conditional and then we don't really need "ospath".  just let normpath do it (or not).

Comment (CaselIT) [reply alembic/testing/util.py]: I'll look if it can be used. If had some problem with the leadit / on linux. But I can update the test to suppt it

**Comment by CaselIT** [comment]
replaced ospath with os.path.normpath

**Comment by sqla-tester** [comment]
New Gerrit review created for change c6b7794510b5179f6ea574019df79a040e3c5f42: https://gerrit.sqlalchemy.org/#/c/sqlalchemy/alembic/+/1627

**Comment by sqla-tester** [comment]
**mike bayer** (zzzeek) wrote:

Code-Review+2 Workflow+1

**Comment by sqla-tester** [comment]
Gerrit review https://gerrit.sqlalchemy.org/1627 has been **merged**. Congratulations! :)
